### PR TITLE
feat: adds l1-l2 how to guide

### PR DESCRIPTION
**Pull request by uF4No**
# Description

Adds a how-to guide on how to send L1->L2 transactions

## Linked Issues

N/A

### CODE REVIEW

**Review by Romsters — APPROVED** [review]

File: content/tutorials/how-to-send-l1-l2-transaction/10.index.md [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+        stateMutability: "nonpayable",
+        type: "function",
+      },
+    ];
+
+    // RPC endpoints
+    const L1_RPC_ENDPOINT = "https://rpc2.sepolia.org";
+    const L2_RPC_ENDPOINT = "https://sepolia.era.zksync.dev";
+
+    const WALLET_PRIV_KEY = process.env.WALLET_PRIV_KEY || "";
```
Comment (jennyg0) [inline content/tutorials/how-to-send-l1-l2-transaction/10.index.md:233]: this doesn't match the .env file `WALLET_PRIVATE_KEY=0x..;`

File: content/tutorials/how-to-send-l1-l2-transaction/10.index.md [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+    ```
+
+    ```bash [yarn]
+    yarn global add ts-node
+
+    ```
+
+    ::
+
+1. In the root folder, add a `.env` file with the private key of the wallet to use:
```
Comment (jennyg0) [inline content/tutorials/how-to-send-l1-l2-transaction/10.index.md:83]: should we add warning to only use test wallet in .env

Comment (uF4No) [reply content/tutorials/how-to-send-l1-l2-transaction/10.index.md:83]: fixed

File: content/tutorials/how-to-send-l1-l2-transaction/10.index.md (hunk 1 above)
Comment (uF4No) [reply content/tutorials/how-to-send-l1-l2-transaction/10.index.md:238]: fixed

File: content/tutorials/how-to-send-l1-l2-transaction/10.index.md [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+    ```ts
+    // retrieve L1 gas price
+    const l1GasPrice = await l1provider.getGasPrice();
+    console.log(`L1 gasPrice ${ethers.utils.formatEther(l1GasPrice)} ETH`);
+    ```
+
+1. Populate the transaction that you want to execute on L2:
+
+    ```ts
+    const contract = new Contract(L2_CONTRACT_ADDRESS, ABI, wallet);
```
Comment (itsacoyote) [inline content/tutorials/how-to-send-l1-l2-transaction/10.index.md:117]:
```suggestion
    const contract = new Contract(<L2_CONTRACT_ADDRESS>, <ABI>, wallet);
```

File: content/tutorials/how-to-send-l1-l2-transaction/10.index.md [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+    // Estimate gas limit for L1-L2 tx
+    const l2GasLimit = await l2provider.estimateGasL1(tx);
+    console.log(`L2 gasLimit ${l2GasLimit.toString()}`);
+    ```
+
+1. Calculate the total transaction fee to cover the cost of sending the
+  transaction on L1 and executing the transaction on L2 using
+  `getBaseCost` from the `Wallet` class. This method:
+
+   - Retrieves the ZKsync BridgeHub contract address calling `zks_getBridgehubContract` RPC method.
```
Comment (itsacoyote) [inline content/tutorials/how-to-send-l1-l2-transaction/10.index.md:135]:
```suggestion
   - Retrieves the ZKsync BridgeHub contract address by calling `zks_getBridgehubContract` RPC method.
```

File: content/tutorials/how-to-send-l1-l2-transaction/10.index.md [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+      gasPrice: l1GasPrice,
+    });
+
+    console.log(`Executing this transaction will cost ${ethers.utils.formatEther(baseCost)} ETH`);
+    ```
+
+1. Encode the transaction calldata:
+
+    ```ts
+    const iface = new ethers.utils.Interface(ABI);
```
Comment (itsacoyote) [inline content/tutorials/how-to-send-l1-l2-transaction/10.index.md:152]:
```suggestion
    const iface = new ethers.utils.Interface(<ABI>);
```

File: content/tutorials/how-to-send-l1-l2-transaction/10.index.md [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+    const l1GasPrice = await l1provider.getGasPrice();
+    console.log(`L1 gasPrice ${ethers.utils.formatEther(l1GasPrice)} ETH`);
+    ```
+
+1. Populate the transaction that you want to execute on L2:
+
+    ```ts
+    const contract = new Contract(L2_CONTRACT_ADDRESS, ABI, wallet);
+    // populate tx object
+    const tx = await contract.populateTransaction.setGreeting(message);
```
Comment (itsacoyote) [inline content/tutorials/how-to-send-l1-l2-transaction/10.index.md:119]: Where is the `message` variable defined?

File: content/tutorials/how-to-send-l1-l2-transaction/10.index.md [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+    });
+
+    console.log(`Executing this transaction will cost ${ethers.utils.formatEther(baseCost)} ETH`);
+    ```
+
+1. Encode the transaction calldata:
+
+    ```ts
+    const iface = new ethers.utils.Interface(ABI);
+    const calldata = iface.encodeFunctionData("setGreeting", [message]);
```
Comment (itsacoyote) [inline content/tutorials/how-to-send-l1-l2-transaction/10.index.md:153]: Where is `message`?

File: content/tutorials/how-to-send-l1-l2-transaction/10.index.md [hunk 8]
```diff
@@ -0,0 +1,340 @@
...
+1. Finally, send the transaction from your wallet using the `requestExecute` method. This helper method:
+
+   - Retrieves the Zksync BridgeHub contract address calling `zks_getBridgehubContract`.
+   - Populates the L1-L2 transaction.
+   - Sends the transaction to the `requestL2TransactionDirect` function of the ZKsync BridgeHub system contract on L1.
+
+    ```ts
+    const txReceipt = await wallet.requestExecute({
+      // destination contract in L2
+      contractAddress: L2_CONTRACT_ADDRESS,
```
Comment (itsacoyote) [inline content/tutorials/how-to-send-l1-l2-transaction/10.index.md:165]:
```suggestion
      contractAddress: <L2_CONTRACT_ADDRESS>,
```

File: content/tutorials/how-to-send-l1-l2-transaction/10.index.md [hunk 9]
```diff
@@ -0,0 +1,340 @@
...
+        stateMutability: "nonpayable",
+        type: "function",
+      },
+    ];
+
+    // RPC endpoints
+    const L1_RPC_ENDPOINT = "https://rpc2.sepolia.org";
+    const L2_RPC_ENDPOINT = "https://sepolia.era.zksync.dev";
+
+    const WALLET_PRIV_KEY = process.env.WALLET_PRIV_KEY || "";
```
Comment (itsacoyote) [inline content/tutorials/how-to-send-l1-l2-transaction/10.index.md:238]:
```suggestion
    const WALLET_PRIV_KEY = process.env.WALLET_PRIVATE_KEY || "";
```

File: content/tutorials/how-to-send-l1-l2-transaction/10.index.md [hunk 10]
```diff
@@ -0,0 +1,340 @@
...
+        value: baseCost,
+        gasPrice: l1GasPrice,
+      },
+    });
+    txReceipt.wait();
+    ```
+
+## Full example
+
+Create a `send-l1-l2-tx.ts` file in the root directory with the next script:
```
Comment (itsacoyote) [inline content/tutorials/how-to-send-l1-l2-transaction/10.index.md:180]:
```suggestion
Create a `send-l1-l2-tx.ts` file in the root directory with the following script:
```

File: content/tutorials/how-to-send-l1-l2-transaction/10.index.md [hunk 11]
```diff
@@ -0,0 +1,340 @@
...
+      txReceipt.wait(1);
+    }
+
+    main()
+      .then()
+      .catch((error) => {
+        console.error(error);
+        process.exitCode = 1;
+      });
+  ::
```
Comment (itsacoyote) [inline content/tutorials/how-to-send-l1-l2-transaction/10.index.md:317]:
```
main()
      .then()
      .catch((error) => {
        console.error(error);
        process.exitCode = 1;
      });
\``` /* backticks missing
::
```
I think the backticks to close the code sample is missing somewhere.

**Review by itsacoyote — CHANGES_REQUESTED** [review]

File: content/tutorials/how-to-send-l1-l2-transaction/10.index.md (hunk 6 above)
Comment (uF4No) [reply content/tutorials/how-to-send-l1-l2-transaction/10.index.md:119]: will add it to a previous step where it's first used

File: content/tutorials/how-to-send-l1-l2-transaction/10.index.md (hunk 7 above)
Comment (uF4No) [reply content/tutorials/how-to-send-l1-l2-transaction/10.index.md:153]: will add in previous step

File: content/tutorials/how-to-send-l1-l2-transaction/10.index.md [hunk 12]
```diff
@@ -0,0 +1,340 @@
...
+## Step-by-step
+
+Sending an L1->L2 transaction requires following these steps:
+
+1. Initialise the providers and the wallet that will send the transaction:
+
+    ```ts
+    const l1provider = new Provider(L1_RPC_ENDPOINT);
+    const l2provider = new Provider(L2_RPC_ENDPOINT);
+    const wallet = new Wallet(WALLET_PRIVATE_KEY, l2provider, l1provider);
```
[the remark this answers is not in the document]
Comment (uF4No) [reply content/tutorials/how-to-send-l1-l2-transaction/10.index.md]: fixed

File: content/tutorials/how-to-send-l1-l2-transaction/10.index.md (hunk 3 above)
Comment (uF4No) [reply content/tutorials/how-to-send-l1-l2-transaction/10.index.md]: fixed

File: content/tutorials/how-to-send-l1-l2-transaction/10.index.md (hunk 5 above)
Comment (uF4No) [reply content/tutorials/how-to-send-l1-l2-transaction/10.index.md]: fixed

File: content/tutorials/how-to-send-l1-l2-transaction/10.index.md (hunk 8 above)
Comment (uF4No) [reply content/tutorials/how-to-send-l1-l2-transaction/10.index.md]: fixed

File: content/tutorials/how-to-send-l1-l2-transaction/10.index.md [hunk 13]
```diff
@@ -0,0 +1,325 @@
...
+- Custom bridges.
+- Multi-layer governing smart contracts.
+
+## Requirements
+
+Before continuing, make sure you have the following:
+
+- Destination contract ABI: in order to execute a transaction on a contract on L2,
+  you'll need the contract's ABI. For this example we'll assume a `Greeter.sol` contract is deployed
+  on ZKsync ([like this one](https://sepolia.explorer.zksync.io/address/0x543A5fBE705d040EFD63D9095054558FB4498F88)).
```
Comment (itsacoyote) [inline content/tutorials/how-to-send-l1-l2-transaction/10.index.md:25]:
```suggestion
  you'll need the contract's ABI. For this example we'll use a `Greeter.sol` contract that is deployed
  on [ZKsync Sepolia testnet](https://sepolia.explorer.zksync.io/address/0x543A5fBE705d040EFD63D9095054558FB4498F88)).
```

**Review by itsacoyote — CHANGES_REQUESTED** [review]
Just one tiny change and everything else looks good 👍🏻

**Review by itsacoyote — APPROVED** [review]

**Pull request merged**